src/chunk.c

```diff
@@ -880,6 +1029,177 @@ chunk_find_all_oids(Hyperspace *hs, List *dimension_vecs, LOCKMODE lockmode)
+
+		/* switch to memory context appropriate for multiple function calls */
+		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+
+		Cache	   *hypertable_cache;
+		Hypertable *ht;
+		Dimension  *time_dim;
+
+		/*
+		 * contains the list of hypertables which need to be considred. this
+		 * is a list containing a single hypertable if PG_ARGISNULL(0) is
+		 * false. Otherwise, it will have the list of all hypertables in the
+		 * system
+		 */
+		List	   *hypertables = NIL;
+		DimensionVec *slices;
+
+		ListCell   *lc;
+		Oid			time_dim_type = InvalidOid;
+
+
+		Oid			table_relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+
+		/*
+		 * get_fn_expr_argtype defaults to UNKNOWNOID if argument is NULL but
+		 * making it InvalidOid makes the logic simpler later
+		 */
+		Oid			older_than_type = PG_ARGISNULL(1) ? InvalidOid : get_fn_expr_argtype(fcinfo->flinfo, 1);
+		Oid			newer_than_type = PG_ARGISNULL(2) ? InvalidOid : get_fn_expr_argtype(fcinfo->flinfo, 2);
+
+		caller_name = PG_NARGS() <= 3 || PG_ARGISNULL(3) ? "show_chunks" : NameStr(*PG_GETARG_NAME(3));
+
+		if (PG_ARGISNULL(3))
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("When calling the internal function for show_chunks "
+							"caller_name cannot be null")
+					 ));
+
+		int64		older_than = -1;
+		int64		newer_than = -1;
+		StrategyNumber start_strategy = InvalidStrategy;
+		StrategyNumber end_strategy = InvalidStrategy;
+
+		if (older_than_type != InvalidOid &&
+			newer_than_type != InvalidOid &&
+			older_than_type != newer_than_type)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("older_than_type and newer_than_type should have the same type")
+					 ));
+
+		/*
+		 * Cache outside the if block to make sure cached hypertable entry
+		 * returned will still be valid in foreach block below
+		 */
+		hypertable_cache = hypertable_cache_pin();
+		if (PG_ARGISNULL(0))
+		{
+			hypertable_get_all(&hypertables, CurrentMemoryContext);
+		}
+		else
+		{
+			ht = hypertable_cache_get_entry(hypertable_cache, table_relid);
+			if (!ht)
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("The given table does not exist or is not a hypertable")
+						 ));
+			hypertables = list_make1(ht);
+		}
+
+		foreach(lc, hypertables)
+		{
+			ht = lfirst(lc);
+
+			/*
+			 * Q:: what is the last argument ( n )of
+			 * hyperspace_get_open_dimension for? looks like it makes the
+			 * function return nth time dimension, but is there not supposed
+			 * to be only one ? It is always called with 0 argument at that
+			 * position.
+			 */
+			time_dim = hyperspace_get_open_dimension(ht->space, 0);
+			if (time_dim == NULL)
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("no hypertables found")
+						 ));
+			if (time_dim_type == InvalidOid)
+				time_dim_type = time_dim->fd.column_type;
+
+			if (time_dim_type != time_dim->fd.column_type)
```

Comment:
Can all this type checking stuff be factored out into a separate function?

Reply:
I mean this until line 1177

Reply:
Ok, as we agreed before, it is much isimpler to factor out the logic into a separate function.
Instead of creating  a new function I moved everything into the function that was supposed to get chunks in range, so instead of giving DimensionSlice vector, I pass range endpoints and types.
It is not a whole lot different or better but this just goes to show that much of what is done is just typechecking which should probably should not be further broken down into functions.
Let me know what you think.